### ISSUE TITLE
0.7.3 Linux CLI binary crashes

### ISSUE THREAD

**Issue by myspotontheweb** [keptn/keptn#2808]
Your issue may already be reported! Please search on the [issue tracker](https://github.com/keptn/keptn/issues) before creating a new one.

**Environment**

* Client OS (e.g., Linux, OSX, Windows): Linux
* Keptn Version (`keptn version`):  0.7.3
* Kubernetes Cloud Provider (e.g., GKE, AKS): minikube
* Kubernetes version (`kubectl version`):

```
$ kubectl version --short
Client Version: v1.18.7
Server Version: v1.16.15
```

**Affected Component**

* [ ] Docs
* [x] CLI
* [ ] Kubernetes Integration
* [ ] Openshift Integration
* [ ] Helm
* [ ] Istio
* [ ] Bridge
* [ ] Gatekeeper
* [ ] Datastore
* [ ] REST API
* [ ] eventbroker / distributor
* [ ] jmeter

**Describe the bug**

Downloaded binary throws 

```
$ ./keptn
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x137b0c6]

goroutine 1 [running]:
github.com/keptn/go-utils/pkg/api/utils.(*APIHandler).GetMetadata(0xc0000ce4b0, 0x0, 0x0)
	/Users/travis/gopath/pkg/mod/github.com/keptn/go-utils@v0.7.2-alpha/pkg/api/utils/apiUtils.go:154 +0x106
github.com/keptn/keptn/cli/cmd.getKeptnServerVersion(0xc00079bea8, 0x1d27b54, 0x5, 0x1)
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/version.go:155 +0x16f
github.com/keptn/keptn/cli/cmd.runDailyVersionCheck()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:118 +0xf2
github.com/keptn/keptn/cli/cmd.Execute()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:47 +0x39
main.main()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/main.go:25 +0x6e
```

I'm running this on Ubuntu

```
$ uname -a
Linux REDACTED 4.15.0-128-generic #131-Ubuntu SMP Wed Dec 9 06:57:35 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**To Reproduce**

```
curl -LO https://github.com/keptn/keptn/releases/download/0.7.3/0.7.3_keptn-linux.tar
tar zxvf 0.7.3_keptn-linux.tar
./keptn
```

**Expected behavior**

I expected the binary to run.

**Comment by christian-kreuzberger-dtx** [comment]
I just tried this on a fresh Linux, and it works for me.

I believe the error you are experiencing is due to the fact that Keptn is trying to connect to a previous cluster that doesn't exist anymore.

The workaround should be to disable the daily version check for now:
```console
keptn set config AutomaticVersionCheck false
```

Nevertheless, I think this is a bug that we need to fix.

**Comment by myspotontheweb** [comment]
I tried disabling the version check, but it's a chick-n-egg issue...  :-(

```
$ rm -rf /home/mark/.keptn
$ ./keptn set config AutomaticVersionCheck false
keptn creates the folder /home/mark/.keptn/ to store logs and possibly creds.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x137b0c6]

goroutine 1 [running]:
github.com/keptn/go-utils/pkg/api/utils.(*APIHandler).GetMetadata(0xc0000cb130, 0x0, 0x0)
	/Users/travis/gopath/pkg/mod/github.com/keptn/go-utils@v0.7.2-alpha/pkg/api/utils/apiUtils.go:154 +0x106
github.com/keptn/keptn/cli/cmd.getKeptnServerVersion(0xc000793ea8, 0x1d27b54, 0x5, 0x1)
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/version.go:155 +0x16f
github.com/keptn/keptn/cli/cmd.runDailyVersionCheck()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:118 +0xf2
github.com/keptn/keptn/cli/cmd.Execute()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:47 +0x39
main.main()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/main.go:25 +0x6e
```

Compiled the CLI from source and it works as expected

```
$ keptn version
keptn creates the folder /home/mark/.keptn/ to store logs and possibly creds.
Kube context has been changed to minikube

Keptn CLI version: develop
Keptn cluster version: Error connecting to server: Head "": unsupported protocol scheme ""
Possible reasons:
* The Keptn API server is currently not available. Check if your Kubernetes cluster is available.
* Your Keptn CLI points to the wrong API server (verify using 'keptn status')

Daily version check is enabled. 
Keptn will collect statistical data and will notify about new versions and security patches for Keptn. Details can be found at: https://keptn.sh/docs/0.7.x/reference/version_check
---------------------------------------------------
* To disable the daily version check, please execute:
 - keptn set config AutomaticVersionCheck false

```

**Comment by christian-kreuzberger-dtx** [comment]
Is it possible that the kube context is set to a cluster that's currently offline? 

Try running
```
kubectl get pods --all-namespaces
```
and if it fails, you can delete the current context.

1. Take a look at the available and current context:
```
kubectl config get-contexts
```

And delete those if they are not working:
```
kubectl config delete-context xyz
```

**Comment by johannes-b** [comment]
I would propose that we investigate this issue when Keptn 0.8.0.alpha is released.

**Comment by myspotontheweb** [comment]
> I believe the error you are experiencing is due to the fact that Keptn is trying to connect to a previous cluster that doesn't exist anymore.

Remove all contexts

```
rm -rf ~/.kube
```

Create new minikube cluster

```
$ minikube start --driver=docker
😄  minikube v1.15.1 on Ubuntu 18.04
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🔥  Creating docker container (CPUs=2, Memory=3900MB) ...
🐳  Preparing Kubernetes v1.19.4 on Docker 19.03.13 ...
🔎  Verifying Kubernetes components...
🌟  Enabled addons: storage-provisioner, default-storageclass
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default
``` 

Download new keptn binary

```
curl -LO https://github.com/keptn/keptn/releases/download/0.7.3/0.7.3_keptn-linux.tar
tar zxvf 0.7.3_keptn-linux.tar
```

Reproduce the problem :-(

```
$ ./keptn
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x137b0c6]

goroutine 1 [running]:
github.com/keptn/go-utils/pkg/api/utils.(*APIHandler).GetMetadata(0xc00080e2d0, 0x0, 0x0)
	/Users/travis/gopath/pkg/mod/github.com/keptn/go-utils@v0.7.2-alpha/pkg/api/utils/apiUtils.go:154 +0x106
github.com/keptn/keptn/cli/cmd.getKeptnServerVersion(0xc00067dea8, 0x1d27b54, 0x5, 0x1)
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/version.go:155 +0x16f
github.com/keptn/keptn/cli/cmd.runDailyVersionCheck()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:118 +0xf2
github.com/keptn/keptn/cli/cmd.Execute()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:47 +0x39
main.main()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/main.go:25 +0x6e
```

kubectl client has no issue

```
$ kubectl get pods -A 
NAMESPACE     NAME                               READY   STATUS    RESTARTS   AGE
kube-system   coredns-f9fd979d6-xhzjb            1/1     Running   0          2m47s
kube-system   etcd-minikube                      1/1     Running   0          2m45s
kube-system   kube-apiserver-minikube            1/1     Running   0          2m45s
kube-system   kube-controller-manager-minikube   1/1     Running   0          2m45s
kube-system   kube-proxy-2bxmn                   1/1     Running   0          2m47s
kube-system   kube-scheduler-minikube            1/1     Running   0          2m45s
kube-system   storage-provisioner                1/1     Running   0          2m51s
```

**Comment by myspotontheweb** [comment]
This may not help......

Downloaded file is statically linked

```
$ file ./keptn
./keptn: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, not stripped
```

Whereas when I compile the binary locally I get a binary that is dynamically linked.

```
$ file /usr/local/bin/keptn
/usr/local/bin/keptn: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=401084030664184f1c6c3f5eaed8c0ca460552b1, not stripped
```

**Comment by christian-kreuzberger-dtx** [comment]
So, looking at the relevant code
https://github.com/keptn/go-utils/blob/b186504cff75c5ffd8f0d84cb70cf3ae7df220c3/pkg/api/utils/apiUtils.go#L153-L155

I'm wondering why this fails...

**Comment by jetzlstorfer** [comment]
we are seeing this issue also in another environment.
any updates on this?

**Comment by christian-kreuzberger-dtx** [comment]
I was unable to reproduce this so far, so any data that you can collect is helpful, especially:
* `keptn version`
* Optional: `md5sum` of keptn binary (just to be 100% sure)
* OS version (e.g., `uname -ar` and `cat /etc/issue`)
* `kubectl version`

**Comment by ekarlso** [comment]
```
md5 of keptn f1b530f53be1958825d2579f943ad1ce

lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.10
Release:	20.10
Codename:	groovy

uname -ar
Linux ekarlso-p1 5.8.0-33-generic #36-Ubuntu SMP Wed Dec 9 09:14:40 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

keptn version: It crashes 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x137b0c6]

goroutine 1 [running]:
github.com/keptn/go-utils/pkg/api/utils.(*APIHandler).GetMetadata(0xc0009ec320, 0x0, 0x0)
	/Users/travis/gopath/pkg/mod/github.com/keptn/go-utils@v0.7.2-alpha/pkg/api/utils/apiUtils.go:154 +0x106
github.com/keptn/keptn/cli/cmd.getKeptnServerVersion(0xc000803ea8, 0x1d27b54, 0x5, 0x1)
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/version.go:155 +0x16f
github.com/keptn/keptn/cli/cmd.runDailyVersionCheck()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:118 +0xf2
github.com/keptn/keptn/cli/cmd.Execute()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:47 +0x39
main.main()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/main.go:25 +0x6e
```

**Comment by christian-kreuzberger-dtx** [comment]
I just tried to reproduce, but no luck. This is what I did:

* Run ubuntu:20.10 docker container: `docker run --rm -it ubuntu:20.10 bash`
* Update package index: `apt update`
* Install curl and sudo: `apt install curl sudo`
* [Install Keptn using quickstart instructions](https://keptn.sh/docs/quickstart/#2-1-install-keptn-cli): `curl -sL https://get.keptn.sh | sudo -E bash`
* Verify the md5sum of keptn: `md5sum $(which keptn)` (FYI: `f1b530f53be1958825d2579f943ad1ce`)
* Run `keptn` aswell as `keptn version`:
   ```
  Keptn CLI version: 0.7.3
  Keptn cluster version: This command requires to be authenticated. See "keptn auth" for details
  
  Daily version check is enabled. 
  Keptn will collect statistical data and will notify about new versions and security patches for Keptn. Details can be found at: https://keptn.sh/docs/0.7.x/reference/version_check
  ---------------------------------------------------
  * To disable the daily version check, please execute:
   - keptn set config AutomaticVersionCheck false
  ```
* [Install kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) using `curl -LO "https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl"`, `chmod +x ./kubectl` and `sudo mv ./kubectl /usr/local/bin/kubectl`
* Run `keptn` as well as `keptn version` again - still works
* Delete the `~/.keptn` folder to ensure everything is fresh: `rm -rf ~/.keptn` 
* Run `keptn` as well as `keptn version` again - still works

**Comment by jetzlstorfer** [comment]
what is really strange: we could successfully do a `keptn auth` in the environment of @ekarlso but the `keptn version` fails.

**Comment by christian-kreuzberger-dtx** [comment]
Now I was able to reproduce this. What's happening:
```
	req.Header.Set("Content-Type", "application/json")
```
fails, as `req` is nil. `req` is nil, because this failed:
```
	req, err := http.NewRequest("GET", s.Scheme+"://"+s.getBaseURL()+"/v1/metadata", nil)
```

And that one failed, because `s.Scheme+"://"+s.getBaseURL()+"/v1/metadata` is not a vaild URL, e.g.:
```
$ echo "http%2f%2ffoobar
foobar" > ~/.keptn/.keptn
$ keptn
```
this results in the same error message:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x137b0c6]

goroutine 1 [running]:
github.com/keptn/go-utils/pkg/api/utils.(*APIHandler).GetMetadata(0xc000141db0, 0x0, 0x0)
	/Users/travis/gopath/pkg/mod/github.com/keptn/go-utils@v0.7.2-alpha/pkg/api/utils/apiUtils.go:154 +0x106
github.com/keptn/keptn/cli/cmd.getKeptnServerVersion(0xc0006a7ea8, 0x1d27b54, 0x5, 0x1)
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/version.go:155 +0x16f
github.com/keptn/keptn/cli/cmd.runDailyVersionCheck()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:118 +0xf2
github.com/keptn/keptn/cli/cmd.Execute()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/cmd/root.go:47 +0x39
main.main()
	/Users/travis/gopath/src/github.com/keptn/keptn/cli/main.go:25 +0x6e
```

So what does this mean? I believe that `keptn auth` was either passed an invalid URL, or `keptn auth` stored the URL in an invalid way in the config file.

**Comment by jetzlstorfer** [comment]
in the config file in `~/.keptn/.keptn` the URL is stored in plain text. no escaping.

**Comment by christian-kreuzberger-dtx** [comment]
Okay, further information: I just installed `gpg` and `pass` as detailed by @ankitjain28may in slack:
```
sudo apt-get install gpg pass -y
gpg --generate-key
pass init <my-public-key>
```
and removed the .keptn directory:
```
rm -rf ~/.keptn
```
After executing
```
keptn
```
I also get the same error/stacktrace from above.

So this definately has something to do with the password store

**Comment by jetzlstorfer** [comment]
this should have been fixed here: https://github.com/keptn/keptn/issues/2638 
can we try with this binary if it still happens?

**Comment by christian-kreuzberger-dtx** [comment]
I can confirm that the current master build does not have this issue.

However, the current master build is most likely not compatible with Keptn 0.7.3.

A workaround (for now) would be to rename `~/.password-store/` to something else  (e.g., `~/.password-store-backup/`), but I'm not sure what this would cause for the password manager on Linux, so **I can not recommend doing that**.

As an alternative workaround, I recommend creating a new user on Linux and working with that user for now, e.g.:
```console
sudo adduser <enter-user-name-here>
sudo -u <enter-user-name-here> bash
```

**Comment by christian-kreuzberger-dtx** [comment]
I have provided a "known issue" and workarounds in https://github.com/keptn/keptn/issues/2808.
As we have already fixed this for 0.8.0, I will close this issue. We currently do not plan to backport it.
